### PR TITLE
Generating can be started manually

### PR DESCRIPTION
**Pull request by matthias-ronge**
Generating images can be started manually. For this purpose, corresponding action links are displayed in the task. If you click on one of the links, the generation will be triggered.

Note: This branch is stacked on [matthias-ronge/KIT-1009](https://github.com/matthias-ronge/kitodo-production/tree/KIT-1009) and should be merged after pull request #1778.

### CODE REVIEW

File: Kitodo/pom.xml [hunk 1]
```diff
@@ -73,6 +73,13 @@
             <groupId>com.sun.jersey</groupId>
             <artifactId>jersey-json</artifactId>
             <version>${jersey.version}</version>
+            <exclusions>
+                <!-- See https://stackoverflow.com/a/42513439/1503237 -->
```
Comment (Kathrin-Huber) [inline Kitodo/pom.xml]: Describe the exclusion, or remove the comment

File: Kitodo/src/main/java/org/kitodo/forms/CurrentTaskForm.java [hunk 2]
```diff
@@ -429,6 +432,46 @@ private void download() {
         }
     }
 
+    /**
+     * Regenerate all images.
+     */
+    public void regenerateAllImagesButtonClick() {
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/forms/CurrentTaskForm.java]: Why the ButtonClick appendix?

File: Kitodo/src/main/java/org/kitodo/tasks/ImageGeneratorStep.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * Enumerates the steps to go to generate images.
+ */
+public enum ImageGeneratorStep {
+    /**
+     * Second step, (if demanded) check if the image exists in the destination
+     * folder, and optionally validate the image file content for not being
+     * corrupted.
+     */
+    DETERMINE_WHICH_IMAGES_NEED_TO_BE_GENERATED,
```
Comment (Kathrin-Huber) [inline Kitodo/src/main/java/org/kitodo/tasks/ImageGeneratorStep.java]: please sort by rank not by alphabet. You wouldn't make an ENUM Month {April, August, December, Februray, January }, would you?

**Comment by Kathrin-Huber** [comment]
Please don't name your modules "SomethingSomethingModule" (skip the Module appendix) there are conventions you can notice from other modules.

File: .gitignore [hunk 4]
```diff
@@ -8,3 +8,4 @@
 /javadoc
 # local build of docs with mkdocs
 /site
+/bin/
```
Comment (Kathrin-Huber) [inline .gitignore]: Why did you add the bin folder?

File: Kitodo/src/main/java/org/kitodo/forms/CurrentTaskForm.java (hunk 2 above)
Comment (matthias-ronge) [reply Kitodo/src/main/java/org/kitodo/forms/CurrentTaskForm.java]: Because it is the method to deal with the event that `regenerateAllImagesButton` is clicked. The method responds to the click event, by adding a long running task to the task manager, which then generates the images. The method does not generate the images itself, so naming it `regenerateAllImages()` would be wrong.

Comment (Beacze) [reply Kitodo/src/main/java/org/kitodo/forms/CurrentTaskForm.java]: We never add ButtonClick to method names and in the form classes most of them are activated by button click....

File: Kitodo-DataManagement/src/main/java/org/kitodo/util/Service.java [hunk 5]
```diff
@@ -9,14 +9,27 @@
...
+
+import javax.imageio.ImageIO;
+
+import org.kitodo.api.filemanagement.FileManagementInterface;
+import org.kitodo.api.imagemanagement.ImageFileFormat;
+import org.kitodo.api.imagemanagement.ImageManagementInterface;
 import org.kitodo.data.database.beans.Folder;
+import org.kitodo.serviceloader.KitodoServiceLoader;
 
 /**
```
[the remark this answers is not in the document]
Comment (Beacze) [reply Kitodo-DataManagement/src/main/java/org/kitodo/util/Service.java]: I would first ask why here KitodoServiceLoader is added to import. According to project design Data Management module can't access API module.

File: Kitodo-DataManagement/src/main/java/org/kitodo/util/Service.java [hunk 6]
```diff
@@ -9,14 +9,27 @@
  * GPL3-License.txt file that was distributed with this source code.
  */
 
-package org.kitodo.forms;
+package org.kitodo.util;
```
[the remark this answers is not in the document]
Comment (Beacze) [reply Kitodo-DataManagement/src/main/java/org/kitodo/util/Service.java]: Package org.kitodo.services is in Core module, this one here is added as new package to Data Management module. Even more, this class is used inside the bean class.

File: Kitodo/src/main/java/org/kitodo/forms/CurrentTaskForm.java (hunk 2 above)
Comment (matthias-ronge) [reply Kitodo/src/main/java/org/kitodo/forms/CurrentTaskForm.java]: We are still missing proper form classes. There is a ticket for this and an intranet page where I documented a suggestion for standardization: https://intranet.slub-dresden.de/display/KITPROD/JSF+Formularklassen In short, the method name is the component name + the event it responds to. Here `Button` is part of the component name `regenerateAllImagesButton` and `Click` is the event. If there was a method that deals with drag and drop onto the button (end drag event), it would be called `regenerateAllImagesButtonEndDrag(...)`

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Folder.java [hunk 7]
```diff
@@ -29,8 +42,13 @@
 import javax.persistence.Table;
 import javax.persistence.Transient;
 
+import org.apache.myfaces.util.FilenameUtils;
+import org.kitodo.api.filemanagement.FileManagementInterface;
 import org.kitodo.api.imagemanagement.ImageManagementInterface;
-import org.kitodo.forms.FolderGenerator;
+import org.kitodo.config.Config;
+import org.kitodo.helper.LexicographicalOrder;
+import org.kitodo.serviceloader.KitodoServiceLoader;
```
Comment (Kathrin-Huber) [inline Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Folder.java]: How can you use the serviceloader from the DataManagement module?
As I see, the kitodo-api was added by you to the dataManagement. this is NOT intendet, as the dataManagement should be independent.

Comment (matthias-ronge) [reply Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Folder.java]: This requires the data management being removed out of the form classes, which did not yet happen. This is inheritance of the legacy code.

File: Kitodo-DataManagement/src/main/java/org/kitodo/util/Service.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (Beacze) [reply Kitodo-DataManagement/src/main/java/org/kitodo/util/Service.java]: I'm quiet confused. The whole functionality of generation could be implemented outside of Data Management module and Task bean used only for setting. Before this PR Task bean was used correctly as a bean, this PR changes it.

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Folder.java (hunk 7 above)
Comment (Beacze) [reply Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Folder.java]: Data Management will not be removed out of form classes as it carries bean objects.... Those functionalities which you implement here, shouldn't be implemented here. Design decision made 2 years ago was that this module is only for data objects, no logic inside.

File: Kitodo-DataManagement/src/main/java/org/kitodo/util/Service.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (matthias-ronge) [reply Kitodo-DataManagement/src/main/java/org/kitodo/util/Service.java]: Unluckily, not. It was “secretly” used in `CurrentTaskForm` and is directly accessed from JSF files. This did not show up as long as there was a 1:1 correspondence between the `Task` classes’ fields and the input field in the JSF to edit a task. But now there is no such correspondence any more, because one task can generate several folders (0−n). So I need a JSF form bean (the `TaskGenerator` class, now renamed to `Generator`) which is repeatable _n_ times, so I need a separate class for being able to repeat it. I do not see a better idea to do so. Correctly, there should be a separate JSF form bean for the task, but we don’t have yet.

File: Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Folder.java (hunk 7 above)
Comment (matthias-ronge) [reply Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Folder.java]: But implementation does not yet follow this decision. So what shell I do? This should have been rewritten with the new front-end, but it wasn’t. If you have a solution for this, let me know.

Comment (Beacze) [reply Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Folder.java]: It was following until you broke it for implementation UGH API (https://github.com/kitodo/kitodo-production/pull/1288). We allowed to make API import until UGH is not removed - right after API dependency will be removed from Data Management module. 

This PR can't use this ServiceLoader, as it is not allowed to add any additional logic to Data Management module.

Comment (matthias-ronge) [reply Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Folder.java]: Do you have a solution that is ok as long as front-end classes are still missing?

Comment (Beacze) [reply Kitodo-DataManagement/src/main/java/org/kitodo/data/database/beans/Folder.java]: Implement it in Core, Task bean takes only get/set.

**Comment by matthias-ronge** [comment]
I will close this pull request as this branch will undergo a complete reworking and needs a new round of review afterwards.
